refactor clientID/clientSecret to be read from GCE metadata; add personal access token support

rjeczalik opened the pull request:
GitHub allows for up to 60 API calls per hour for unauthenticated requests.
If GITHUB_TOKEN is exported to the environemnt prior to starting gddo-server,
each request will use it for token authentication. For more details see
https://developer.github.com/v3/#rate-limiting.

Closes #254.

rjeczalik commented:
/cc @adg @garyburd PTAL

garyburd commented:
What is the reason for moving the code to set the credentials on a request from the transport to package level variables in the gosrc package?

rjeczalik commented:
@garyburd None actually, I'll move those to the transport.

rjeczalik commented:
@garyburd I took the opportunity and simplified passing contactEmail and githubCreds to lintapp and talksapp - the former used config.go and the latter secret.json - right now both use env_variables defined in the app.yaml file. Moved also configuring github auth to the transport for gddo-server.

What do you think?

garyburd commented:
Nice job. LGTM.

adg commented:
This looks good to me, but I'm not going to merge it until I have occasion to update the corresponding deployment scripts. Thanks for the work.

garyburd commented:
To avoid accidentally publishing private repos on the godoc.org, the server should not make authenticated calls to GitHub. The server should instead use a [client id and secret to raise the rate limit](https://developer.github.com/v3/#increasing-the-unauthenticated-rate-limit-for-oauth-applications). This is the mechanism used before this PR.

rjeczalik commented:
@garyburd I'd rather say it's misuse of the token access feature - if one grants private repo access for a token, then sure, gddo-server will display private repos of the repo the token is attached to. If some does not want private repos to show up, it's as easy as removing "repo" from the scopes.

Does the risk of misusing the tokens warrant adding more code to bring back the clientID/secret? IMHO it's redundant.

garyburd commented:
Client id and secret are sufficient. There's no need for token. The difference in code is

    r.Header.Set("Authorization", "token "+githubToken)

for token and 

    if r.URL.RawQuery == "" {
        r.URL.RawQuery = gitHubCredentials
    } else {
        r.URL.RawQuery += "&" + gitHubCredentials
    }

for  client id and secret.

garyburd commented:
As always, this is @adg's call.  Now that he's running the service, any error will expose his repos instead of mine.

adg commented:
I would prefer to use the approach we were already using.

On 1 October 2015 at 06:26, Gary Burd <notifications@github.com> wrote:

> As always, this is @adg <https://github.com/adg>'s call. Now that he's
> running the service, any error will expose his repos instead of mine.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/golang/gddo/pull/330#issuecomment-144531391>.
>

rjeczalik commented:
@adg Ack, do you want me replace the token with clientid/secret and keep the refactoring or abandon the cl?

adg commented:
Does it still satisfy your aim if you keep the refactoring? Or does this entirely defeat what you set out to achieve with this CC?

rjeczalik commented:
@adg Ideally I'd keep both - having the clientid/secret and token, and when both are specified - the former preferred. The use-case I have is a cluster with a couple of services that already use github token, so I could spare on devops not needing to maintain the clientid/secret as well. I use gddo-server internally to display docs for private projects defaulting for osx platform as oppose to linux.

I understand that having more code (token support) that no-one will use is more cost than feature so I have no problem with just keeping the refactoring :)

adg commented:
Well let's support both then. I don't mind. It's not much code.

garyburd commented:
Oh, I was confused about the goal of this PR. Because you talked about limiting access in the issue and client id / secret is already limited, I thought the important feature here was using environment variables. I apologize for the confusion.

Replacing config.go with environment variables and metadata service seems like a good thing to me.

rjeczalik commented:
@garyburd Oh sorry for confusion, I'll update the commit's description as well.

rjeczalik commented:
Hey @garyburd @adg I've brought back the clientID/clientSecret creds, they've been refactored accordingly to the token one to be read from:

- `github-client-id` and `github-client-secret` GCE metadata attributes
- `GITHUB_CLIENT_ID` and `GITHUB_CLIENT_SECRET` env vars

Please take a look.

adg commented on gddo-server/client.go:
```diff
@@ -47,32 +48,15 @@ func timeoutDial(network, addr string) (net.Conn, error) {
+
+var httpClient = &http.Client{
+	Transport: transport,
+	Timeout:   *requestTimeout,
```
This will use the value that requestTimeout points to at init time, before flags are parsed. Probably not what you want.

adg commented on gddo-server/main.go:
```diff
@@ -814,18 +816,33 @@ var (
+func init() {
+	if metadata.OnGCE() {
+		clientID, err := metadata.ProjectAttributeValue("github-client-id")
+		if err == nil && clientID != "" {
```
if err != nil, log it. we'll want to know why.

adg commented on gddo-server/main.go:
```diff
@@ -814,18 +816,33 @@ var (
 )
 
+func init() {
+	if metadata.OnGCE() {
```
why not have a `NewTransportFromMetadata` function, just like `NewTransportFromEnvironment` ? seems clearer

adg commented on httputil/transport.go:
```diff
@@ -0,0 +1,59 @@
+package httputil
```
needs a copyright notice; see the other files

adg commented on lintapp/README.md:
```diff
@@ -6,6 +6,6 @@ This directory contains the source for [go-lint.appspot.com](http://go-lint.apps
-- Run the server using the dev_appserver command.
+- Edit env\_variables in the `app.yaml` file as described in the comments.
+- Install Go App Engine SDK.
+- Run the server using the dev\_appserver command.
```
is the underscore really necessary? that's a shame
maybe

`dev_appserver.py`

adg commented on talksapp/README.md:
```diff
@@ -6,7 +6,7 @@ This directory contains the source for [go-talks.appspot.com](http://go-talks.ap
 -----------------------------
 
-- Copy config.go.template to config.go and edit the file as described in the comments.
+- Edit env\_variables in the `app.yaml` file as described in the comments.
```
have this say "make a copy of app.yaml named prod.yaml and put in the authentication data". we don't want these values accidentally committed and leaked

rjeczalik commented:
@adg All addressed, please take a look.

adg commented on gddo-server/main.go:
```diff
@@ -809,28 +809,29 @@ func defaultBase(path string) string {
+	firstGetTimeout = flag.Duration("first_get_timeout", 5*time.Second, "Time to wait for first fetch of package from the VCS.")
+	maxAge          = flag.Duration("max_age", 24*time.Hour, "Update package documents older than this age.")
+	httpAddr        = flag.String("http", ":8080", "Listen for HTTP connections on this address.")
+	defaultGOOS     = flag.String("default_goos", "", "Default GOOS to use when building package documents.")
```
Why has the order of these changed?

adg commented on httputil/transport.go:
```diff
@@ -0,0 +1,97 @@
+}
+
+func (t *Transport) RoundTrip(req *http.Request) (*http.Response, error) {
+	if t.UserAgent != "" {
```
RoundTrippers aren't supposed to mutate the provided request.
What you need to do here is something like `cloneRequest` in this file https://github.com/golang/oauth2/blob/master/transport.go#L96 before you mutate the headers.

adg commented on httputil/transport.go:
```diff
@@ -0,0 +1,97 @@
+	"google.golang.org/cloud/compute/metadata"
+)
+
+type Transport struct {
```
This should be called `AuthTransport` and have a doc comment

```
// AuthTransport is an implementation of http.RoundTripper that authenticates with the GitHub API.

rjeczalik commented:
@adg Another round, PTAL.

adg commented:
It looks good, I'm just holding off on deploying it until I have time to make the corresponding production/deployment changes.

rjeczalik commented:
Rebased + cleaned code a bit in fe6d289.

adg commented:
This looks good, and I've tested that it works. Can you please squash the commits into a single commit? Thanks.

rjeczalik commented:
@adg: Hey, rebased with master and squashed into single commit.

adg commented:
Thanks!

adg commented:
@campoy can you please update the talks app? It'll need updates to its app.yaml file. (Also, do you maintain the lint app? Is that even deployed somewhere? (@garyburd?))

adg commented:
This is now deployed to godoc.org

shurcooL commented:
> Is that even deployed somewhere?

Lint app is deployed at http://go-lint.appspot.com/.

jehiah commented:
@rjeczalik thanks for wrapping this up; I've been successfully using gddo at Bitly for the past few months (thanks to this PR) against a private GitHub repo :star: :star2: :rocket:

rjeczalik commented:
Hey @jehiah nice to hear that :-)

garyburd commented:
The lint app is running at  http://go-lint.appspot.com/ and is owned by @campoy.  There are links to the lint app from the godoc tool pages ([example](https://godoc.org/github.com/golang/gddo/gosrc?tools)).
